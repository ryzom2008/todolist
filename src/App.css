.container {
    width: 90%;
    max-width: 1024px;
    margin: 0 auto;
    display: flex;
    justify-content: space-between;
    height: 200px;
    padding: 0 20px;
    background-color: #EBF5FB;
}

.gamesList {
    min-width: 200px;
}

.score {
    display: flex;
}

.score-item {
    margin: 0 20px;
    text-align: center;
}

.actions {
    text-align: center;
    margin-top: 20px;
}

.button {
    width: 60px;
    margin: 0 20px;
    justify-content: center;
}