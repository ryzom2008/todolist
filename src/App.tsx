import React, {useState} from 'react';
import './App.css';
import {Game, Score} from "./index";

type Props = {
    games: Game[];
    scores: Score[];
}

const App = ({games, scores}:Props) => {
    const [upcomingGames, setUpcomingGames] = useState<Game[]>(games);
    const [finishedGames, setFinishedGames] = useState<Game[]>([]);
    const [currentGame, setCurrentGame] = useState<Game | null>(null);
    const startGame = () => {
        const [nextGame, ...restGames] = upcomingGames;
        setCurrentGame(nextGame as Game);
        setUpcomingGames(restGames);
    }
    const onScoreUpdate = () => {
        const currentGameScore = scores.find(el => el.gameId === currentGame?.id);
        if (currentGameScore) {
            const updatedGame = {
                ...currentGame,
                home: {...currentGame?.home, score: currentGameScore.home},
                away: {...currentGame?.away, score: currentGameScore.away}
            }
            setCurrentGame(updatedGame as Game);
        }
    }
    const onGameFinish = () => {
        const sortedFinishedGames = [currentGame as Game, ...finishedGames]
            .sort((g1, g2) => (g2.home.score + g2.away.score) - (g1.home.score + g1.away.score));
        setFinishedGames(sortedFinishedGames);
        setCurrentGame(null);
    }
    return (
        <div className="container">
            <div className="gamesList">
                <h3>Upcoming games</h3>
                {upcomingGames.map(el => <div
                    data-testid="upcoming-game"
                    key={el.id}>{el.home.country} - {el.away.country}</div>)}
            </div>
            <div>
                <div className="score">
                    <div className="score-item">
                        <h4>home team</h4>
                        <span data-testid="home-team">{currentGame?.home.country}</span>
                    </div>
                    <div className="score-item">
                        <h4>score</h4>
                        <span data-testid="score">{currentGame?.home.score || 0} - {currentGame?.away.score || 0}</span>
                    </div>
                    <div className="score-item">
                        <h4>away team</h4>
                        <span data-testid="away-team">{currentGame?.away.country}</span>
                    </div>
                </div>
                <div className="actions">
                    {currentGame &&
                    <>
                      <button className="button" onClick={onScoreUpdate}>Update</button>
                      <button className="button" onClick={onGameFinish}>Finish</button>
                    </>}
                    {!currentGame && !!upcomingGames.length &&
                    <button className="button" onClick={startGame}>Start</button>
                    }
                    {!upcomingGames.length && !currentGame && <div>No upcoming games</div>}
                </div>
            </div>
            <div className="gamesList">
                <h3>Finished games</h3>
                {finishedGames.map(el => <div
                    data-testid="finished-game"
                    key={el.id}>{el.home.country} {el.home.score} - {el.away.country} {el.away.score}</div>)}
            </div>
        </div>
    );
}

export default App;
