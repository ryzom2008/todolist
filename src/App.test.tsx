import React from 'react';
import {fireEvent, render, RenderResult} from '@testing-library/react';
import App from './App';
import {Game, Score} from "./index";

const testGames: Game[] = [
    {id: 1, home: {country: 'Mexico', score: 0}, away: {country: 'Canada', score: 0}},
    {id: 2, home: {country: 'Spain', score: 0}, away: {country: 'Brazil', score: 0}},
    {id: 3, home: {country: 'Germany', score: 0}, away: {country: 'France', score: 0}},
];

const testScores: Score[] = [
    {gameId: 1, home: 0, away: 5},
    {gameId: 2, home: 10, away: 2},
    {gameId: 3, home: 6, away: 6},
];

describe("App", () => {
    let component: RenderResult;
    beforeEach(() =>{
        component = render(<App games={testGames} scores={testScores}/>)
    });

    test('should render component correctly', () => {
        expect(component).toMatchSnapshot();
    });

    test('should set a current game', () => {
        const startButton = component.getByText('Start');
        fireEvent.click(startButton);
        const homeTeamEl = component.getByTestId('home-team');
        const awayTeamEl = component.getByTestId('away-team');
        const scoreEl = component.getByTestId('score');
        const updateButton = component.getByText('Update');
        const finishButton = component.getByText('Finish');
        expect(homeTeamEl.textContent).toEqual('Mexico');
        expect(awayTeamEl.textContent).toEqual('Canada');
        expect(scoreEl.textContent).toEqual('0 - 0');
        expect(startButton).not.toBeInTheDocument();
        expect(updateButton).toBeInTheDocument();
        expect(finishButton).toBeInTheDocument();
    });

    test('should update a current game score', () => {
        const startButton = component.getByText('Start');
        fireEvent.click(startButton);
        const updateButton = component.getByText('Update');
        fireEvent.click(updateButton);
        const scoreEl = component.getByTestId('score');

        expect(scoreEl.textContent).toEqual('0 - 5');
    });

    test('should finish a current game', () => {
        const startButton = component.getByText('Start');
        fireEvent.click(startButton);
        const updateButton = component.getByText('Update');
        const finishButton = component.getByText('Finish');
        const scoreEl = component.getByTestId('score');
        fireEvent.click(updateButton);
        fireEvent.click(finishButton);
        expect(scoreEl.textContent).toEqual('0 - 0');
        const upcomingGames = component.getAllByTestId('upcoming-game').map(el => el.textContent);
        const finishedGames = component.getAllByTestId('finished-game').map(el => el.textContent);
        expect(upcomingGames).toEqual([
            "Spain - Brazil",
            "Germany - France",
        ]);
        expect(updateButton).not.toBeInTheDocument();
        expect(finishButton).not.toBeInTheDocument();
        expect(component.getByText('Start')).toBeInTheDocument();
        expect(finishedGames).toEqual(["Mexico 0 - Canada 5"]);
    });

    test('should show "no upcoming games" message when all the games are finished', () => {
        let upcomingGamesDivs = component.queryAllByTestId('upcoming-game');
        let updateButton, finishButton, startButton;
        while (upcomingGamesDivs.length) {
            startButton = component.queryByText('Start');
            fireEvent.click(startButton as HTMLElement);
            updateButton = component.getByText('Update');
            finishButton = component.getByText('Finish');
            fireEvent.click(updateButton);
            fireEvent.click(finishButton);
            upcomingGamesDivs = component.queryAllByTestId('upcoming-game');
        }
        const upcomingGames = upcomingGamesDivs.map(el => el.textContent);
        const finishedGames = component.getAllByTestId('finished-game').map(el => el.textContent);
        expect(upcomingGames).toEqual([]);
        expect(updateButton).not.toBeInTheDocument();
        expect(finishButton).not.toBeInTheDocument();
        expect(startButton).not.toBeInTheDocument();
        expect(finishedGames).toEqual([
            "Germany 6 - France 6",
            "Spain 10 - Brazil 2",
            "Mexico 0 - Canada 5",
        ]);
        expect(component.getByText("No upcoming games")).toBeInTheDocument();
    })
});
