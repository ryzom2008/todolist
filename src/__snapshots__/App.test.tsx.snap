// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="gamesList"
        >
          <h3>
            Upcoming games
          </h3>
          <div
            data-testid="upcoming-game"
          >
            Mexico
             - 
            Canada
          </div>
          <div
            data-testid="upcoming-game"
          >
            Spain
             - 
            Brazil
          </div>
          <div
            data-testid="upcoming-game"
          >
            Germany
             - 
            France
          </div>
        </div>
        <div>
          <div
            class="score"
          >
            <div
              class="score-item"
            >
              <h4>
                home team
              </h4>
              <span
                data-testid="home-team"
              />
            </div>
            <div
              class="score-item"
            >
              <h4>
                score
              </h4>
              <span
                data-testid="score"
              >
                0
                 - 
                0
              </span>
            </div>
            <div
              class="score-item"
            >
              <h4>
                away team
              </h4>
              <span
                data-testid="away-team"
              />
            </div>
          </div>
          <div
            class="actions"
          >
            <button
              class="button"
            >
              Start
            </button>
          </div>
        </div>
        <div
          class="gamesList"
        >
          <h3>
            Finished games
          </h3>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="gamesList"
      >
        <h3>
          Upcoming games
        </h3>
        <div
          data-testid="upcoming-game"
        >
          Mexico
           - 
          Canada
        </div>
        <div
          data-testid="upcoming-game"
        >
          Spain
           - 
          Brazil
        </div>
        <div
          data-testid="upcoming-game"
        >
          Germany
           - 
          France
        </div>
      </div>
      <div>
        <div
          class="score"
        >
          <div
            class="score-item"
          >
            <h4>
              home team
            </h4>
            <span
              data-testid="home-team"
            />
          </div>
          <div
            class="score-item"
          >
            <h4>
              score
            </h4>
            <span
              data-testid="score"
            >
              0
               - 
              0
            </span>
          </div>
          <div
            class="score-item"
          >
            <h4>
              away team
            </h4>
            <span
              data-testid="away-team"
            />
          </div>
        </div>
        <div
          class="actions"
        >
          <button
            class="button"
          >
            Start
          </button>
        </div>
      </div>
      <div
        class="gamesList"
      >
        <h3>
          Finished games
        </h3>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`App should render screen correctly 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`should render screen correctly 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
